src/Sentry/Internal/Http/HttpTransport.cs

```diff
@@ -31,14 +37,66 @@ internal class HttpTransport : ITransport
 
         public async Task SendEnvelopeAsync(Envelope envelope, CancellationToken cancellationToken = default)
         {
-            var request = CreateRequest(envelope);
+            var instant = DateTimeOffset.Now;
+
+            // Discard envelope items if they don't fit in the quota limit
+            var envelopeItems = new List<EnvelopeItem>();
+            foreach (var envelopeItem in envelope.Items)
+            {
+                // Check if there is at least one matching category for this item that is rate-limited
+                var isRateLimited = _categoryLimitResets
+                    .Where(kvp => kvp.Value > instant)
+                    .Any(kvp => kvp.Key.Matches(envelopeItem));
+
+                if (!isRateLimited)
+                {
+                    envelopeItems.Add(envelopeItem);
+                }
+                else
+                {
+                    _options.DiagnosticLogger?.LogDebug(
+                        "Envelope item of type {0} was discarded because it's rate-limited.",
+                        envelopeItem.TryGetType()
+                    );
+                }
+            }
+
+            if (!envelopeItems.Any())
+            {
+                _options.DiagnosticLogger?.LogInfo(
+                    "Envelope {0} was discarded because all contained items are rate-limited.",
+                    envelope.TryGetEventId()
+                );
+                return;
+            }
+
+            // TODO: optimize/refactor
```

Comment:
Do we need this TODO? Maybe elaborate what optimization?

Reply:
Just want to refactor before we merge it. Don't like how this method looks right now.